src/AutoRest.CSharp/Common/Output/Builders/SerializationBuilder.cs

```diff
@@ -310,19 +313,21 @@ public JsonObjectSerialization BuildJsonObjectSerialization(ObjectSchema objectS
         private class PropertyBag
         {
             public Dictionary<string, PropertyBag> Bag { get; } = new Dictionary<string, PropertyBag>();
-            public List<Property> Properties { get; } = new List<Property>();
+            public List<ObjectTypeProperty> Properties { get; } = new List<ObjectTypeProperty>();
         }
 
         private static void PopulatePropertyBag(PropertyBag propertyBag, int depthIndex)
         {
-            foreach (Property property in propertyBag.Properties.ToArray())
+            foreach (ObjectTypeProperty property in propertyBag.Properties.ToArray())
             {
-                if (depthIndex >= (property.FlattenedNames?.Count ?? 0) - 1)
+                var schemaProperty = property.SchemaProperty!; // we ensure this is not null when we build the array
+                ICollection<string> flattenedNames = property.SerializationMapping?.SerializationPath as ICollection<string> ?? schemaProperty.FlattenedNames;
```

Comment:
`property.SerializationMapping?.SerializationPath` is `IReadOnlyList`, casting it to `ICollection` basically is to iterate it and turn it into a non-readonly collection.
Another option is to duplicate the logic:
```csharp
string name;
int serializationPathCount = property.SerializationMapping?.SerializationPath?.Count ?? 0;
if (serializationPathCount > 0) 
{
    if (depthIndex >= serializationPathCount - 1)
    {
        {
            continue;
        }
        name = property.SerializationMapping?.SerializationPath!.ElementAt(depthIndex);
    }
}
else
{
    if (depthIndex >= (flattenedNames?.Count ?? 0) - 1)
    {
        continue;
    }

    name = flattenedNames!.ElementAt(depthIndex);
}
```

Reply:
This makes code harder to read.
The value of `SerializationMapping?.SerializationPath` is meant to override the existing things in `FlattenedNames` if configured.
We have to have the cast because their types are not the same, and I do not want to call `ToArray` or `ToList` because that will allocate another two collections.

Reply:
The code does get more complex by duplication, which does not make any copy and no iteration. 
I can take either here.